scripts/nmt/encoder_decoder.py

```diff
@@ -584,6 +595,599 @@ def hybrid_forward(self, F, step_input, states):  #pylint: disable=arguments-dif
         return rnn_out, new_states, step_additional_outputs
 
 
+def position_encoding_init(max_length, dim):
+    ''' Init the sinusoid position encoding table '''
+    position_enc = np.arange(max_length).reshape((-1, 1)) \
+                   / (np.power(10000, (2. / dim) * np.arange(dim).reshape((1, -1))))
+    # Apply the cosine to even columns and sin to odds.
+    position_enc[:, 0::2] = np.sin(position_enc[:, 0::2])  # dim 2i
+    position_enc[:, 1::2] = np.cos(position_enc[:, 1::2])  # dim 2i+1
+    return position_enc
+
+
+class PositionwiseFFN(HybridBlock):
+    """Structure of the Positionwise Feed-Forward Neural Network.
+
+    Parameters
+    ----------
+    units : int
+    hidden_size : int
+        number of units in the hidden layer of position-wise feed-forward networks
+    dropout : float
+    use_residual : bool
+    weight_initializer : str or Initializer
+        Initializer for the input weights matrix, used for the linear
+        transformation of the inputs.
+    bias_initializer : str or Initializer
+        Initializer for the bias vector.
+    activation : str, default 'relu'
+    	Activation function
+    prefix : str, default 'rnn_'
+        Prefix for name of `Block`s
+        (and name of weight if params is `None`).
+    params : Parameter or None
+        Container for weight sharing between cells.
+        Created if `None`.
+    """
+    def __init__(self, units=512, hidden_size=2048, dropout=0.0, use_residual=True,
+                 weight_initializer=None, bias_initializer='zeros', activation='relu',
+                 prefix=None, params=None):
+        super(PositionwiseFFN, self).__init__(prefix=prefix, params=params)
+        self._hidden_size = hidden_size
+        self._units = units
+        self._use_residual = use_residual
+        self.ffn_1 = nn.Dense(units=hidden_size, flatten=False, activation=activation,
+                              weight_initializer=weight_initializer, bias_initializer=bias_initializer,
+                              prefix='ffn_1_')
+        self.ffn_2 = nn.Dense(units=units, flatten=False,
+                              weight_initializer=weight_initializer, bias_initializer=bias_initializer,
+                              prefix='ffn_2_')
+        self.dropout_layer = nn.Dropout(dropout)
+        self.layer_norm = nn.LayerNorm()
```

Comment:
We'd better add a prefix for layernorm. Also, do we need to use `with self.name_scope():`?